### ISSUE TITLE
Run dials.export format=mmcif compress=bz2 automatically

### ISSUE THREAD

**Issue by dagewa** [xia2/xia2#534]
As [promised](https://github.com/dials/dials/issues/1457#issuecomment-721816985) I'm creating an issue here about writing out the deposition file-of-record (mmcif) from xia2 runs. With https://github.com/dials/dials/pull/1482 merged this could be compressed to save disk space.

**Comment by jbeilstenedmands** [comment]
I'm going to start working on this now, unless anyone else has already started doing so in secret 🙂

**Comment by dagewa** [comment]
I was waiting for dials/dials#1482 to go in first, but there isn't really any reason to hold off. You know better where to look in xia2 than I do, so you're welcome to it ;-)

One thing I did want to try was to compare the filesizes of unmerged MTZ, uncompressed mmcif and the two supported compression modes for mmcif for some reasonably large unmerged case. Perhaps also the speed of writing the compressed files. Although the title suggests to use `gz`, the `bz2` file might be significantly smaller.

**Comment by dagewa** [comment]
An experiment with a fairly meaty dataset (from `/dls/metrix/metrix/data_analysis/for_METRIX/results_20190326/xia2_stresstest/4L0N/`) shows a few interesting things:
```
fcx32934@zdar:/tmp$ time dials.export scaled.expt scaled.refl > /dev/null

real	0m16.622s
user	0m10.640s
sys	0m5.725s
fcx32934@zdar:/tmp$ time dials.export scaled.expt scaled.refl format=mmcif > /dev/null

real	2m37.118s
user	2m29.220s
sys	0m7.862s
fcx32934@zdar:/tmp$ time dials.export scaled.expt scaled.refl format=mmcif compress=gz > /dev/null

real	4m14.879s
user	4m7.603s
sys	0m7.121s
fcx32934@zdar:/tmp$ time dials.export scaled.expt scaled.refl format=mmcif compress=bz2 > /dev/null

real	2m48.300s
user	2m42.135s
sys	0m6.161s
fcx32934@zdar:/tmp$ ll -h scaled.mtz scaled.cif scaled.cif.gz scaled.cif.bz2
-rw-r--r-- 1 fcx32934 fcx32934 280M Nov  6 17:15 scaled.cif
-rw-r--r-- 1 fcx32934 fcx32934  87M Nov  6 17:23 scaled.cif.bz2
-rw-r--r-- 1 fcx32934 fcx32934 103M Nov  6 17:20 scaled.cif.gz
-rw-r--r-- 1 fcx32934 fcx32934 129M Nov  6 17:12 scaled.mtz
```
The mmcif output is _really slow_ even before asking for compression. The bz2 output was, to my surprise, faster than gz. As I saw before, uncompressed unmerged mmcif is a bit over twice the size of the MTZ.

**Comment by dagewa** [comment]
Here's a repeat of the above test after https://github.com/dials/dials/pull/1464 was merged:
```
$ time dials.export scaled.expt scaled.refl > /dev/null

real	0m18.642s
user	0m11.688s
sys	0m6.270s
fcx32934@zdar:/tmp$ time dials.export scaled.expt scaled.refl format=mmcif > /dev/null

real	2m6.911s
user	2m1.089s
sys	0m5.802s
fcx32934@zdar:/tmp$ time dials.export scaled.expt scaled.refl format=mmcif compress=gz > /dev/null

real	3m28.851s
user	3m23.284s
sys	0m5.561s
fcx32934@zdar:/tmp$ time dials.export scaled.expt scaled.refl format=mmcif compress=bz2 > /dev/null

real	2m16.603s
user	2m10.920s
sys	0m5.680s
```

so it's a little bit quicker, but still rather slow. However, the files are now different in size:
```
fcx32934@zdar:/tmp$ ll -h scaled.mtz scaled.cif scaled.cif.gz scaled.cif.bz2
-rw-r--r-- 1 fcx32934 fcx32934 191M Nov  9 09:11 scaled.cif
-rw-r--r-- 1 fcx32934 fcx32934  38M Nov  9 09:16 scaled.cif.bz2
-rw-r--r-- 1 fcx32934 fcx32934  47M Nov  9 09:14 scaled.cif.gz
-rw-r--r-- 1 fcx32934 fcx32934 129M Nov  9 09:09 scaled.mtz
```
I will try a before/after file comparison

**Comment by dagewa** [comment]
Ok, the difference in format is like this. Before:
```
...
loop_
  _pdbx_diffrn_unmerged_refln.reflection_id
  _pdbx_diffrn_unmerged_refln.scan_id
  _pdbx_diffrn_unmerged_refln.image_id_begin
  _pdbx_diffrn_unmerged_refln.image_id_end
  _pdbx_diffrn_unmerged_refln.index_h
  _pdbx_diffrn_unmerged_refln.index_k
  _pdbx_diffrn_unmerged_refln.index_l
  _pdbx_diffrn_unmerged_refln.scale_value
  _pdbx_diffrn_unmerged_refln.intensity_meas
  _pdbx_diffrn_unmerged_refln.intensity_sigma
  _pdbx_diffrn_unmerged_refln.scan_angle_reflection
  _pdbx_diffrn_unmerged_refln.partiality
        1  1    0    2   34   168  -41   1.044084646563654       -9.025341938514508    38.86951988380073       -53.88127988124704  0.9078091491439713
...
```
After:
```
...
  _pdbx_diffrn_unmerged_refln.partiality
            1  1    0    2   34   168  -41   1.04408      -9.02534    38.8695      -53.8813  0.907809
...
```
So the new version has a much reduced field width, explaining the smaller file sizes

**Comment by Anthchirp** [comment]
What is the overall xia2 runtime with your example? 2 minutes just to compress a file is rather considerable.
Could you paste the timing report section from near the end of `xia2-debug.txt`?

**Comment by dagewa** [comment]
I didn't run xia2 sorry

**Comment by dagewa** [comment]
however, there was an original xia2 run  at `/dls/metrix/metrix/data_analysis/for_METRIX/results_20190326/xia2_stresstest/4L0N/`. It is not really comparable, because it did not use dials.scale, but aimless and was run on a different machine to the above test. However, to get a feel for timings on this dataset here is that timing summary:
```
Timing summary:
1.6 sec: dials.generate_mask
407.5 sec: dials.find_spots
2.6 sec: dials.detect_blanks
16.4 sec: dials.search_beam_position
35.5 sec: dials.index
37.5 sec: dials.refine_bravais_settings
1.6 sec: dials.reindex
6.3 sec: dials.reindex
15.0 sec: dials.refine
29.2 sec: dials.refine
2293.3 sec: dials.integrate
40.5 sec: dials.export
14.7 sec: pointless
30.1 sec: dials.export
18.8 sec: pointless
9.9 sec: sortmtz
17.8 sec: pointless
17.4 sec: pointless
10.5 sec: sortmtz
171.1 sec: aimless
38.0 sec: xia2.resolutionizer
126.2 sec: aimless
126.1 sec: aimless
44.7 sec: aimless
24.1 sec: dials.reindex
36.3 sec: dials.two_theta_refine
46.5 sec: aimless
50.6 sec: ctruncate
2.2 sec: freerflag
2.1 sec: freerflag

Longest times:
2293.3 sec: dials.integrate
407.5 sec: dials.find_spots
171.1 sec: aimless
126.2 sec: aimless
126.1 sec: aimless
50.6 sec: ctruncate
46.5 sec: aimless
44.7 sec: aimless
40.5 sec: dials.export
38.0 sec: xia2.resolutionizer

```

**Comment by dagewa** [comment]
Following tests in https://github.com/dials/dials/issues/1485 I would recommend `.bz2` compression rather than `.gz`

**Comment by dagewa** [comment]
@CV-GPhL suggests these datasets as simple, single-sweep test cases:
> 5E73 5E9I 6CCU 6CDC 6DEX 6FJ8 6GFN 6GT5

A further list of PAD data:
> 4TWN 4XGU 4YUO 4ZC9 5B8D 5C66 5E73 5E74 5EO9 5HQG 5J1S 5J1T 5KCO 5KH9 5KO9 5KWR 5KWS 5M10 5M4L 5NHU 5OF9 5OFA 5OFB 5SZB 5SZC 5T4M 5UCG 5UK4 5UUS 5V0V 5V0Z 5VBB 5VBD 5VBT 5VM2 5VOG 5VP5 5VZR 5WB0 5WB1 6BHC 6BY0 6BY9 6BYA 6BZG 6C1T 6CDD 6FID 6FJ2 6FJ4 6FJ8 6FJ9

We would like to process these with xia2, produce mmcif and compare results to highlight any inconsistencies in definitions. For example, the majority of PDB entries contain [_reflns_shellmeani_over_sigi_obs](https://gist.github.com/CV-GPhL/21003c9812357dba83128076a818a6a9#item-_reflns_shellmeani_over_sigi_obs), but this is not a useful quantity following its actual definition.

Diffraction images for the above PDB IDs are available. There are links from the PDB page, for example [4TWN](https://www.rcsb.org/structure/4TWN) contains 
> Diffraction Data DOI: 10.15785/SBGRID/93 [SBGrid](https://data.sbgrid.org/dataset/93/)

This might be a useful start point to ensure we are producing correct mmcif files.

**Comment by rjgildea** [comment]
I think this is now resolved by cfb5c51
